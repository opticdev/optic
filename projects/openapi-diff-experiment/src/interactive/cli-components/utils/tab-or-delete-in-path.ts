import { fragmentize } from '../../../services/diff/differs/url-path-diff';
import niceTry from 'nice-try';

export function tabInPath(pattern: string, targetUrl: string) {
  const targetFragments = fragmentize(targetUrl).filter((i) => i !== '');
  const inputFragments = (niceTry(() => fragmentize(pattern)) || []).filter(
    (i) => i !== ''
  );

  if (targetFragments.length > inputFragments.length) {
    const newFragment = targetFragments[inputFragments.length];

    return '/' + [...inputFragments, newFragment].join('/');
  }
  return pattern;
}
