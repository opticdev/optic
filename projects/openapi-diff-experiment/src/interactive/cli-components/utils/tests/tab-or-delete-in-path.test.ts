import { tabInPath } from '../tab-or-delete-in-path';

describe('tab or delete in path', () => {
  it('will append last component', () => {
    expect(
      tabInPath('/example/{exampleId}', '/example/123/status')
    ).toMatchSnapshot();
  });
  it('will append last component when trailing slash', () => {
    expect(
      tabInPath('/example/{exampleId}/', '/example/123/status')
    ).toMatchSnapshot();
  });

  it('will do nothing when full path', () => {
    expect(
      tabInPath('/example/{exampleId}/status', '/example/123/status')
    ).toMatchSnapshot();
  });

  it('will append constant 2nd component', () => {
    expect(tabInPath('/example', '/example/123/status')).toMatchSnapshot();
  });
});
