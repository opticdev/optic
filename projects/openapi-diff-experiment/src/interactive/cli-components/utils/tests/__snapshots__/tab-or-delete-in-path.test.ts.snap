// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tab or delete in path will append constant 2nd component 1`] = `"/example/123"`;

exports[`tab or delete in path will append last component 1`] = `"/example/{exampleId}/status"`;

exports[`tab or delete in path will append last component when trailing slash 1`] = `"/example/{exampleId}/status"`;

exports[`tab or delete in path will do nothing when full path 1`] = `"/example/{exampleId}/status"`;
